server.go

```diff
@@ -0,0 +1,221 @@
+package goconnpool
+
+import (
+	"context"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/cenkalti/backoff"
+	"github.com/pkg/errors"
+)
+
+type connectionProvider interface {
+	getConnection(ctx context.Context) (Conn, error)
+	retryTimeout() time.Duration
+}
+
+type server struct {
+	mu sync.Mutex
+
+	addr           string
+	connectTimeout time.Duration
+
+	nOpenedConns int
+	maxConns     int
+	openedConns  deck
+
+	reqDuration time.Duration
+	lastUsage   time.Time
+
+	dialer Dialer
+
+	bOff        backoff.BackOff
+	nextBackoff time.Time
+	down        bool
+
+	clock Clock
+}
+
+func newServerWrapper(addr string, cfg Config) connectionProvider {
+	return newServer(addr, cfg)
+}
+
+func newServer(addr string, cfg Config) *server {
+	bc := backoff.NewExponentialBackOff()
+	bc.InitialInterval = cfg.InitialBackoffInterval
+	bc.MaxInterval = cfg.MaxBackoffInterval
+	bc.MaxElapsedTime = 0
+	bc.Clock = cfg.Clock
+
+	bc.Reset() // required to re-setup config options
+
+	if cfg.backoffRandomizationFactor != nil {
+		// only for tests. Default backoff interval should be used in production
+		bc.RandomizationFactor = *cfg.backoffRandomizationFactor
+	}
+
+	return &server{
+		addr:     addr,
+		maxConns: cfg.MaxConnsPerServer,
+		dialer:   cfg.Dialer,
+		bOff:     bc,
+
+		connectTimeout: cfg.ConnectTimeout,
+
+		reqDuration: time.Duration(1000000.0/float64(cfg.MaxRPS)) * time.Microsecond,
+
+		clock: cfg.Clock,
+	}
+}
+
+func (s *server) updateLastUsage() bool {
+	if s.getRatelimitTimeout() > 0 {
+		return false
+	}
+
+	s.lastUsage = s.clock.Now()
```

Comment:
You dont use mutex here because you dont care about real last usage timestamp, yeah?

Reply:
`updateLastUsage` is called from mutex opened in the previous stack-frame: `getConnection` call.

Reply:
hm, its not really obvious.

Reply:
I've add a comment =)